[Crash] placeOrder with Invalid pairID

Current Behavior: placeOrder with Invalid pairID causes XUD to crash, due to unhandled gRPC error
How it should be: gracefully handle all gRPC errors

```
2018-7-30 13:37:14 [RPC] error: invalid pairId: BTC/LTC
TypeError: Incorrectly typed arguments to startBatch
    at handleError (/home/kilrau/xud/node_modules/grpc/src/server.js:71:8)
    at sendUnaryData (/home/kilrau/xud/node_modules/grpc/src/server.js:597:9)
    at GrpcService.<anonymous> (/home/kilrau/xud/dist/grpc/GrpcService.js:40:17)
    at Generator.throw (<anonymous>)
    at rejected (/home/kilrau/xud/dist/grpc/GrpcService.js:5:65)
    at <anonymous>
    at runMicrotasksCallback (internal/process/next_tick.js:121:5)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
/home/kilrau/xud/dist/bootstrap.js:13
        throw err;
        ^

TypeError: Incorrectly typed arguments to startBatch
    at handleError (/home/kilrau/xud/node_modules/grpc/src/server.js:71:8)
    at sendUnaryData (/home/kilrau/xud/node_modules/grpc/src/server.js:597:9)
    at GrpcService.<anonymous> (/home/kilrau/xud/dist/grpc/GrpcService.js:40:17)
    at Generator.throw (<anonymous>)
    at rejected (/home/kilrau/xud/dist/grpc/GrpcService.js:5:65)
    at <anonymous>
    at runMicrotasksCallback (internal/process/next_tick.js:121:5)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! xud@1.0.0-alpha.1 start: `node dist/Xud.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the xud@1.0.0-alpha.1 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/kilrau/.npm/_logs/2018-07-30T20_37_14_899Z-debug.log
```

[2018-07-30T20_37_14_899Z-debug.log](https://github.com/ExchangeUnion/xud/files/2242807/2018-07-30T20_37_14_899Z-debug.log)

--- a/lib/cli/command.ts
+++ b/lib/cli/command.ts
@@ -20,7 +20,7 @@ interface grpcResponse {
  */
 export const callback = (error: Error | null, response: grpcResponse) => {
   if (error) {
-    console.error(error);
+    console.error(`${error.name}: ${error.message}`);
   } else {
     console.log(JSON.stringify(response.toObject(), null, 2));
   }

--- a/lib/constants/errorCodesPrefix.ts
+++ b/lib/constants/errorCodesPrefix.ts
@@ -4,4 +4,5 @@ export default {
   ORDERBOOK: '3',
   LND: '4',
   RAIDEN: '5',
+  SERVICE: '6',
 };

--- a/lib/grpc/GrpcService.ts
+++ b/lib/grpc/GrpcService.ts
@@ -1,4 +1,4 @@
-import grpc from 'grpc';
+import grpc, { status, ServiceError } from 'grpc';
 import Logger from '../Logger';
 import Service from '../service/Service';
 import { isObject } from '../utils/utils';
@@ -8,6 +8,8 @@ import { GetInfoResponse } from '../proto/lndrpc_pb';
 import { Orders } from 'lib/orderbook/OrderBook';
 import { MatchingResult } from '../types/matchingEngine';
 import { OwnOrder } from '../types/orders';
+import { default as orderErrors, errorCodes as orderErrorCodes } from '../orderbook/errors';
+import { default as serviceErrors, errorCodes as serviceErrorCodes } from '../service/errors';
 
 function serializeDateProperties(response) {
   Object.keys(response).forEach((key) => {
@@ -36,8 +38,20 @@ class GrpcService {
       const response = serializeDateProperties(rawResponse);
       callback(null, response);
     } catch (err) {
+      // if we recognize this error, return a proper gRPC ServiceError with a descriptive and appropriate code
+      let grpcError: ServiceError | undefined;
+      switch (err.code) {
+        case serviceErrorCodes.INVALID_ARGUMENT:
+          grpcError = { ...err, code: status.INVALID_ARGUMENT };
+          break;
+        case orderErrorCodes.INVALID_PAIR_ID:
+          grpcError = { ...err, code: status.NOT_FOUND };
+          break;
+      }
       this.logger.error(err);
-      callback(err, null);
+
+      // return grpcError if we've created one, otherwise pass along the caught error as UNKNOWN
+      callback(grpcError ? grpcError : err, null);
     }
   }
 

--- a/lib/lndclient/errors.ts
+++ b/lib/lndclient/errors.ts
@@ -1,11 +1,16 @@
 import errorCodesPrefix from '../constants/errorCodesPrefix';
 
 const codesPrefix = errorCodesPrefix.LND;
-const errors: any = {};
-
-errors.LND_IS_DISABLED = {
-  message: 'lnd is disabled',
-  code: codesPrefix.concat('1'),
+const errorCodes = {
+  LND_IS_DISABLED: codesPrefix.concat('.1'),
 };
 
+const errors = {
+  LND_IS_DISABLED: {
+    message: 'lnd is disabled',
+    code: errorCodes.LND_IS_DISABLED,
+  },
+};
+
+export { errorCodes };
 export default errors;

--- a/lib/orderbook/OrderBook.ts
+++ b/lib/orderbook/OrderBook.ts
@@ -24,6 +24,11 @@ interface OrderBook {
   emit(event: 'peerOrder', order: orders.StampedPeerOrder);
 }
 
+type OrderArrays = {
+  buyOrders: orders.StampedOrder[],
+  sellOrders: orders.StampedOrder[],
+};
+
 class OrderBook extends EventEmitter {
   public pairs: db.PairInstance[] = [];
   public matchingEngines: { [ pairId: string ]: MatchingEngine } = {};
@@ -82,18 +87,18 @@ class OrderBook extends EventEmitter {
   /**
    * Returns lists of buy and sell orders of peers
    */
-  public getPeerOrders = (pairId: string, maxResults: number): { [ type: string ]: orders.StampedPeerOrder[] } => {
-    return this.getOrders(maxResults, this.peerOrders[pairId]) as { [type: string]: orders.StampedPeerOrder[] };
+  public getPeerOrders = (pairId: string, maxResults: number): OrderArrays => {
+    return this.getOrders(maxResults, this.peerOrders[pairId]);
   }
 
   /*
   * Returns lists of the node's own buy and sell orders
   */
-  public getOwnOrders = (pairId: string, maxResults: number): { [type: string]: orders.StampedOwnOrder[] } => {
-    return this.getOrders(maxResults, this.ownOrders[pairId]) as { [type: string]: orders.StampedOwnOrder[] };
+  public getOwnOrders = (pairId: string, maxResults: number): OrderArrays => {
+    return this.getOrders(maxResults, this.ownOrders[pairId]);
   }
 
-  private getOrders = (maxResults: number, orders: Orders): { [type: string]: orders.StampedOrder[] } => {
+  private getOrders = (maxResults: number, orders: Orders): OrderArrays => {
     if (maxResults > 0) {
       return {
         buyOrders: Object.values(orders.buyOrders).slice(0, maxResults),
@@ -263,8 +268,8 @@ class OrderBook extends EventEmitter {
     const promises: Promise<orders.OutgoingOrder | void>[] = [];
     for (const { id } of pairs) {
       const orders = await this.getOwnOrders(id, 0);
-      orders['buyOrders'].forEach(order => promises.push(this.createOutgoingOrder(order)));
-      orders['sellOrders'].forEach(order => promises.push(this.createOutgoingOrder(order)));
+      orders['buyOrders'].forEach(order => promises.push(this.createOutgoingOrder(order as orders.StampedOwnOrder)));
+      orders['sellOrders'].forEach(order => promises.push(this.createOutgoingOrder(order as orders.StampedOwnOrder)));
     }
     await Promise.all(promises).then((outgoingOrders) => {
       peer.sendOrders(outgoingOrders as orders.OutgoingOrder[]);
@@ -310,4 +315,4 @@ class OrderBook extends EventEmitter {
 }
 
 export default OrderBook;
-export { Orders };
+export { Orders, OrderArrays };

--- a/lib/orderbook/errors.ts
+++ b/lib/orderbook/errors.ts
@@ -7,11 +7,11 @@ const errorCodes = {
 };
 
 const errors = {
-  INVALID_PAIR_ID: pairId => ({
+  INVALID_PAIR_ID: (pairId: string) => ({
     message: `invalid pairId: ${pairId}`,
     code: errorCodes.INVALID_PAIR_ID,
   }),
-  DUPLICATED_ORDER: localId => ({
+  DUPLICATED_ORDER: (localId: string) => ({
     message: `duplicated localId: ${localId}`,
     code: codesPrefix.concat('.2'),
   }),

--- a/lib/orderbook/errors.ts
+++ b/lib/orderbook/errors.ts
@@ -1,16 +1,21 @@
 import errorCodesPrefix from '../constants/errorCodesPrefix';
 
 const codesPrefix = errorCodesPrefix.ORDERBOOK;
-const errors: any = {};
+const errorCodes = {
+  INVALID_PAIR_ID: codesPrefix.concat('.1'),
+  DUPLICATED_ORDER: codesPrefix.concat('.2'),
+};
 
-errors.INVALID_PAIR_ID = pairId => ({
-  message: `invalid pairId: ${pairId}`,
-  code: codesPrefix.concat('1'),
-});
+const errors = {
+  INVALID_PAIR_ID: pairId => ({
+    message: `invalid pairId: ${pairId}`,
+    code: errorCodes.INVALID_PAIR_ID,
+  }),
+  DUPLICATED_ORDER: localId => ({
+    message: `duplicated localId: ${localId}`,
+    code: codesPrefix.concat('.2'),
+  }),
+};
 
-errors.DUPLICATED_ORDER = localId => ({
-  message: `duplicated localId: ${localId}`,
-  code: codesPrefix.concat('2'),
-});
-
+export { errorCodes };
 export default errors;

--- a/lib/p2p/errors.ts
+++ b/lib/p2p/errors.ts
@@ -1,8 +1,21 @@
-class P2PError {
-  constructor(public message: string) {}
-}
+import errorCodesPrefix from '../constants/errorCodesPrefix';
 
-export default {
-  ADDRESS_ALREADY_CONNECTED: (address: string) => new P2PError(`Address (${address}) already connected`),
-  NOT_CONNECTED: (address: string) => new P2PError(`Address (${address}) is not connected`),
+const codesPrefix = errorCodesPrefix.P2P;
+const errorCodes = {
+  ADDRESS_ALREADY_CONNECTED: codesPrefix.concat('.1'),
+  NOT_CONNECTED: codesPrefix.concat('.2'),
 };
+
+const errors = {
+  ADDRESS_ALREADY_CONNECTED: (address: string) => ({
+    message: `Address (${address}) already connected`,
+    code: errorCodes.ADDRESS_ALREADY_CONNECTED,
+  }),
+  NOT_CONNECTED: (address: string) => ({
+    message: `Address (${address}) is not connected`,
+    code: errorCodes.NOT_CONNECTED,
+  }),
+};
+
+export { errorCodes };
+export default errors;

--- a/lib/raidenclient/RaidenClient.ts
+++ b/lib/raidenclient/RaidenClient.ts
@@ -2,7 +2,7 @@ import http from 'http';
 
 import Logger from '../Logger';
 import BaseClient, { ClientStatus } from '../BaseClient';
-import * as errors from './errors';
+import errors from './errors';
 
 /**
  * A utility function to parse the payload from an http response.

--- a/lib/raidenclient/errors.ts
+++ b/lib/raidenclient/errors.ts
@@ -1,33 +1,41 @@
 import errorCodesPrefix from '../constants/errorCodesPrefix';
 
 const codesPrefix = errorCodesPrefix.RAIDEN;
-
-export const RAIDEN_IS_DISABLED = {
-  message: 'raiden is disabled',
-  code: codesPrefix.concat('1'),
+const errorCodes = {
+  RAIDEN_IS_DISABLED: codesPrefix.concat('.1'),
+  INSUFFICIENT_BALANCE: codesPrefix.concat('.2'),
+  TIMEOUT: codesPrefix.concat('.3'),
+  INVALID: codesPrefix.concat('.4'),
+  SERVER_ERROR: codesPrefix.concat('.5'),
+  UNEXPECTED: codesPrefix.concat('.6'),
 };
 
-export const INSUFFICIENT_BALANCE = {
-  message: 'insufficient balance to perform request',
-  code: codesPrefix.concat('2'),
+const errors = {
+  RAIDEN_IS_DISABLED: {
+    message: 'raiden is disabled',
+    code: errorCodes.RAIDEN_IS_DISABLED,
+  },
+  INSUFFICIENT_BALANCE: {
+    message: 'insufficient balance to perform request',
+    code: errorCodes.INSUFFICIENT_BALANCE,
+  },
+  TIMEOUT: {
+    message: 'request timed out',
+    code: errorCodes.TIMEOUT,
+  },
+  INVALID: {
+    message: 'request is conflicting or otherwise invalid',
+    code: errorCodes.INVALID,
+  },
+  SERVER_ERROR: {
+    message: 'raiden server error',
+    code: errorCodes.SERVER_ERROR,
+  },
+  UNEXPECTED: {
+    message: 'unexpected error during raiden request',
+    code: errorCodes.UNEXPECTED,
+  },
 };
 
-export const TIMEOUT = {
-  message: 'request timed out',
-  code: codesPrefix.concat('3'),
-};
-
-export const INVALID = {
-  message: 'request is conflicting or otherwise invalid',
-  code: codesPrefix.concat('4'),
-};
-
-export const SERVER_ERROR = {
-  message: 'raiden server error',
-  code: codesPrefix.concat('5'),
-};
-
-export const UNEXPECTED = {
-  message: 'unexpected error during raiden request',
-  code: codesPrefix.concat('6'),
-};
+export { errorCodes };
+export default errors;

--- a/lib/service/Service.ts
+++ b/lib/service/Service.ts
@@ -1,6 +1,6 @@
 import Logger from '../Logger';
 import Pool from '../p2p/Pool';
-import OrderBook from '../orderbook/OrderBook';
+import OrderBook, { OrderArrays } from '../orderbook/OrderBook';
 import LndClient from '../lndclient/LndClient';
 import RaidenClient, { TokenSwapPayload } from '../raidenclient/RaidenClient';
 import { OwnOrder } from '../types/orders';
@@ -144,10 +144,26 @@ class Service extends EventEmitter {
   }
 
   /**
-   * Get a list of standing orders from the order book.
+   * Get a list of standing peer orders from the order book for a specified trading pair, or for all
+   * trading pairs if no pair is specified.
    */
-  public getOrders = ({ pairId, maxResults }: { pairId: string, maxResults: number }) => {
-    return this.orderBook.getPeerOrders(pairId, maxResults);
+  public getOrders = ({ pairId, maxResults }: { pairId?: string, maxResults?: number }) => {
+    const ret: OrderArrays = {
+      buyOrders: [],
+      sellOrders: [],
+    };
+    if (pairId) {
+      const orderArrays = this.orderBook.getPeerOrders(pairId, maxResults ? maxResults : 0);
+      ret.buyOrders.concat(orderArrays.buyOrders);
+      ret.sellOrders.concat(orderArrays.sellOrders);
+    } else {
+      this.orderBook.pairs.forEach((pair) => {
+        const orderArrays = this.orderBook.getPeerOrders(pair.id, maxResults ? maxResults : 0);
+        ret.buyOrders.concat(orderArrays.buyOrders);
+        ret.sellOrders.concat(orderArrays.sellOrders);
+      });
+    }
+    return ret;
   }
 
   /**

--- a/lib/service/Service.ts
+++ b/lib/service/Service.ts
@@ -1,4 +1,3 @@
-import assert from 'assert';
 import Logger from '../Logger';
 import Pool from '../p2p/Pool';
 import OrderBook from '../orderbook/OrderBook';
@@ -7,8 +6,8 @@ import RaidenClient, { TokenSwapPayload } from '../raidenclient/RaidenClient';
 import { OwnOrder } from '../types/orders';
 import Config from '../Config';
 import { EventEmitter } from 'events';
-import { orders } from '../types';
 import SocketAddress from '../p2p/SocketAddress';
+import errors from './errors';
 
 const packageJson = require('../../package.json');
 
@@ -24,6 +23,12 @@ export type ServiceComponents = {
   /** The function to be called to shutdown the parent process */
   shutdown: Function;
 };
+
+function checkArgument(expectedCondition: boolean, message: string) {
+  if (!expectedCondition) {
+    throw errors.INVALID_ARGUMENT(message);
+  }
+}
 
 /** Class containing the available RPC methods for XUD */
 class Service extends EventEmitter {
@@ -149,13 +154,10 @@ class Service extends EventEmitter {
    * Add an order to the order book.
    */
   public placeOrder = async (order: OwnOrder) => {
-    assert(order.price >= 0, 'price cannot be negative');
-    assert(order.quantity !== 0, 'quantity must not equal 0');
-    if (order.price === 0) {
-      return this.orderBook.addMarketOrder(order);
-    } else {
-      return this.orderBook.addLimitOrder(order);
-    }
+    checkArgument(order.price >= 0, 'price cannot be negative');
+    checkArgument(order.quantity !== 0, 'quantity must not equal 0');
+
+    return order.price > 0 ? this.orderBook.addLimitOrder(order) : this.orderBook.addMarketOrder(order);
   }
 
   /*

--- a/lib/service/errors.ts
+++ b/lib/service/errors.ts
@@ -1,0 +1,16 @@
+import errorCodesPrefix from '../constants/errorCodesPrefix';
+
+const codesPrefix = errorCodesPrefix.SERVICE;
+const errorCodes = {
+  INVALID_ARGUMENT: codesPrefix.concat('.1'),
+};
+
+const errors = {
+  INVALID_ARGUMENT: (message: string) => ({
+    message,
+    code: errorCodes.INVALID_ARGUMENT,
+  }),
+};
+
+export { errorCodes };
+export default errors;

